File: js/llm.js
import OpenAI from 'openai';
import Anthropic from '@anthropic-ai/sdk';
import { GoogleGenerativeAI, HarmCategory, HarmBlockThreshold } from '@google/generative-ai';
import Groq from 'groq-sdk';

/**
 * Abstract Base Class for a Large Language Model provider.
 * It defines the interface for interacting with different LLM APIs.
 */
class LLM {
    constructor(apiKey, model) {
        if (this.constructor === LLM) {
            throw new Error("LLM is an abstract class and cannot be instantiated directly.");
        }
        this.apiKey = apiKey;
        this.model = model;
    }

    async chatCompletion(options) {
        throw new Error("Method 'chatCompletion()' must be implemented.");
    }
}

export class OpenAIProvider extends LLM {
    constructor(apiKey = null, model = 'gpt-4o') {
        const resolvedApiKey = apiKey || process.env.OPENAI_API_KEY;
        if (!resolvedApiKey) {
            throw new Error("OPENAI_API_KEY must be provided or set as an environment variable.");
        }
        super(resolvedApiKey, model);
        this.client = new OpenAI({ apiKey: this.apiKey });
    }

    async chatCompletion({ messages, temperature, maxTokens, ...kwargs }) {
        console.log(`Calling OpenAI API with model: ${this.model}`);
        const completion = await this.client.chat.completions.create({
            model: this.model,
            messages,
            temperature,
            max_tokens: maxTokens,
            ...kwargs,
        });
        // Normalize the response to match the Gemini/Claude format.
        return {
            choices: [{ message: { role: 'assistant', content: completion.choices[0].message.content } }]
        };
    }
}

export class XAIProvider extends LLM {
    constructor(apiKey = null, model = 'grok-4') {
        const resolvedApiKey = apiKey || process.env.XAI_API_KEY;
        if (!resolvedApiKey) {
            throw new Error("XAI_API_KEY must be provided or set as an environment variable.");
        }
        super(resolvedApiKey, model);
        this.client = new OpenAI({
            apiKey: this.apiKey,
            baseURL: 'https://api.x.ai/v1',
        });
    }

    async chatCompletion({ messages, temperature, maxTokens, ...kwargs }) {
        console.log(`Calling xAI (Grok) API with model: ${this.model}`);
        const completion = await this.client.chat.completions.create({
            model: this.model,
            messages,
            temperature,
            max_tokens: maxTokens,
            ...kwargs,
        });
        // Normalize the response to match the Gemini/Claude format.
        return completion;
    }
}

export class GroqProvider extends LLM {
    constructor(apiKey = null, model = 'llama3-8b-8192') {
        const resolvedApiKey = apiKey || process.env.GROQ_API_KEY;
        if (!resolvedApiKey) {
            throw new Error("GROQ_API_KEY must be provided or set as an environment variable.");
        }
        super(resolvedApiKey, model);
        this.client = new Groq({ apiKey: this.apiKey });
    }

    async chatCompletion({ messages, temperature, maxTokens, ...kwargs }) {
        console.log(`Calling Groq API with model: ${this.model}`);
        const completion = await this.client.chat.completions.create({
            model: this.model,
            messages,
            temperature,
            max_tokens: maxTokens,
            ...kwargs,
        });
        // Normalize the response to match the Gemini/Claude format.
        return {
            choices: [{ message: { role: 'assistant', content: completion.choices[0].message.content } }]
        };
    }
}

export class ClaudeProvider extends LLM {
    constructor(apiKey = null, model = 'claude-3-opus-20240229') {
        const resolvedApiKey = apiKey || process.env.ANTHROPIC_API_KEY;
        if (!resolvedApiKey) {
            throw new Error("ANTHROPIC_API_KEY must be provided or set as an environment variable.");
        }
        super(resolvedApiKey, model);
        this.client = new Anthropic({ apiKey: this.apiKey });
    }

    async chatCompletion({ messages, temperature, maxTokens, ...kwargs }) {
        console.log(`Calling Anthropic (Claude) API with model: ${this.model}`);
        let systemPrompt = '';
        if (messages.length > 0 && messages[0].role === 'system') {
            systemPrompt = messages.shift().content;
        }

        const completion = await this.client.messages.create({
            model: this.model,
            system: systemPrompt,
            messages: messages,
            temperature,
            max_tokens: maxTokens,
            ...kwargs,
        });
        // Anthropic's response structure is different. We normalize it to look like OpenAI's.
        return {
            choices: [{ message: { role: 'assistant', content: completion.content[0].text } }]
        };
    }
}

export class GeminiProvider extends LLM {
    constructor(apiKey = null, model = 'gemini-1.5-flash') {
        const resolvedApiKey = apiKey || process.env.GOOGLE_API_KEY;
        if (!resolvedApiKey) {
            throw new Error("GOOGLE_API_KEY must be provided or set as an environment variable.");
        }
        super(resolvedApiKey, model);
        this.genAI = new GoogleGenerativeAI(this.apiKey);
    }

    async chatCompletion({ messages, temperature, maxTokens, ...kwargs }) {
        console.log(`Calling Google (Gemini) API with model: ${this.model}`);
        let systemPrompt = '';
        if (messages.length > 0 && messages[0].role === 'system') {
            systemPrompt = messages.shift().content;
        }

        const model = this.genAI.getGenerativeModel({
            model: this.model,
            systemInstruction: systemPrompt,
        });

        // Gemini requires a specific format for history.
        // For simplicity, we'll just use the last user message.
        // A more robust implementation would convert the full history.
        const userPrompt = messages[messages.length - 1].content;

        const result = await model.generateContent(userPrompt, {
            temperature,
            maxOutputTokens: maxTokens,
        });
        const response = await result.response;
        // Gemini's response structure is different. We normalize it to look like OpenAI's.
        return {
            choices: [{ message: { role: 'assistant', content: response.text() } }]
        };
    }
}

const PROVIDERS = {
    openai: OpenAIProvider,
    xai: XAIProvider,
    groq: GroqProvider,
    claude: ClaudeProvider,
    gemini: GeminiProvider,
};

export function getLlm(providerName, options = {}) {
    const ProviderClass = PROVIDERS[providerName.toLowerCase()];
    if (!ProviderClass) {
        throw new Error(`Unknown LLM provider: '${providerName}'. Available: ${Object.keys(PROVIDERS).join(', ')}`);
    }
    return new ProviderClass(options.apiKey, options.model);
}

File: python/llm.py
from abc import ABC, abstractmethod
from typing import List, Dict, Any, Optional, Type
import os
import warnings
import logging

# Each provider has its own library.
# You will need to install them:
# pip install openai anthropic google-generativeai groq
from openai import AsyncOpenAI
import anthropic
import google.generativeai as genai
from groq import AsyncGroq

logger = logging.getLogger(__name__)

class LLM(ABC):
    """
    Abstract Base Class for a Large Language Model provider.
    It defines the interface for interacting with different LLM APIs.
    """

    def __init__(self, api_key: Optional[str] = None, model: Optional[str] = None):
        self.api_key = api_key
        self.model = model

    @abstractmethod
    async def chat_completion(
        self,
        messages: List[Dict[str, str]],
        temperature: float,
        max_tokens: int,
        **kwargs: Any
    ) -> Dict[str, Any]:
        """
        Generates a chat completion response.

        Args:
            messages: A list of message dictionaries, e.g., [{"role": "user", "content": "Hello"}].
            temperature: The sampling temperature for the completion.
            max_tokens: The maximum number of tokens to generate.
            **kwargs: Additional provider-specific arguments.

        Returns:
            A dictionary representing the normalized response from the LLM.
        """
        pass


class OpenAIProvider(LLM):
    """LLM provider for OpenAI models."""

    def __init__(self, api_key: Optional[str] = None, model: str = "gpt-4o"):
        resolved_api_key = api_key or os.getenv("OPENAI_API_KEY")
        if not resolved_api_key:
            raise ValueError("OPENAI_API_KEY must be provided or set as an environment variable.")

        super().__init__(api_key=resolved_api_key, model=model)
        self.client = AsyncOpenAI(api_key=self.api_key)

    async def chat_completion(
        self,
        messages: List[Dict[str, str]],
        temperature: float,
        max_tokens: int,
        **kwargs: Any
    ) -> Dict[str, Any]:
        """Generates a chat completion response using the OpenAI API."""
        # Separate system prompt if present, as it's a top-level parameter for some models
        system_prompt = ""
        if messages and messages[0]['role'] == 'system':
            system_prompt = messages[0]['content']
            messages = messages[1:]

        completion = await self.client.chat.completions.create(
            model=self.model,
            messages=messages,
            temperature=temperature,
            max_tokens=max_tokens,
            **kwargs
        )
        return completion.dict()

class Grok(LLM):
    """LLM provider for xAI (Grok) models, using an OpenAI-compatible API."""

    def __init__(self, api_key: Optional[str] = None, model: str = "grok-4"):
        resolved_api_key = api_key or os.getenv("XAI_API_KEY")
        if not resolved_api_key:
            raise ValueError("XAI_API_KEY must be provided or set as an environment variable.")

        super().__init__(api_key=resolved_api_key, model=model)
        self.client = AsyncOpenAI(
            api_key=self.api_key,
            base_url="https://api.x.ai/v1"
        )

    async def chat_completion(
        self,
        messages: List[Dict[str, str]],
        temperature: float,
        max_tokens: int,
        **kwargs: Any
    ) -> Dict[str, Any]:
        """Generates a chat completion response using the xAI API."""
        logger.info(
            "Calling xAI API.",
            extra={"model": self.model, "temperature": temperature, "max_tokens": max_tokens}
        )
        completion = await self.client.chat.completions.create(
            model=self.model,
            messages=messages,
            temperature=temperature,
            max_tokens=max_tokens,
            **kwargs
        )
        # Normalize the response to a standard dictionary format
        return completion.dict()


class GroqProvider(LLM):
    """LLM provider for Groq models."""

    def __init__(self, api_key: Optional[str] = None, model: str = "llama3-8b-8192"):
        resolved_api_key = api_key or os.getenv("GROQ_API_KEY")
        if not resolved_api_key:
            raise ValueError("GROQ_API_KEY must be provided or set as an environment variable.")

        super().__init__(api_key=resolved_api_key, model=model)
        self.client = AsyncGroq(api_key=self.api_key)

    async def chat_completion(
        self,
        messages: List[Dict[str, str]],
        temperature: float,
        max_tokens: int,
        **kwargs: Any
    ) -> Dict[str, Any]:
        """Generates a chat completion response using the Groq API."""
        logger.info(
            "Calling Groq API.",
            extra={"model": self.model, "temperature": temperature, "max_tokens": max_tokens}
        )
        completion = await self.client.chat.completions.create(
            model=self.model,
            messages=messages,
            temperature=temperature,
            max_tokens=max_tokens,
            **kwargs
        )
        # Normalize the response to a standard dictionary format
        return completion.dict()


class Claude(LLM):
    """LLM provider for Anthropic (Claude) models."""

    def __init__(self, api_key: Optional[str] = None, model: str = "claude-3-opus-20240229"):
        resolved_api_key = api_key or os.getenv("ANTHROPIC_API_KEY")
        if not resolved_api_key:
            raise ValueError("ANTHROPIC_API_KEY must be provided or set as an environment variable.")

        super().__init__(api_key=resolved_api_key, model=model)
        self.client = anthropic.AsyncAnthropic(api_key=self.api_key)

    async def chat_completion(
        self,
        messages: List[Dict[str, str]],
        temperature: float,
        max_tokens: int,
        **kwargs: Any
    ) -> Dict[str, Any]:
        """Generates a chat completion response using the Anthropic API."""
        system_prompt = ""
        if messages and messages[0]['role'] == 'system':
            system_prompt = messages[0]['content']
            messages = messages[1:]

        logger.info(
            "Calling Anthropic API.",
            extra={"model": self.model, "temperature": temperature, "max_tokens": max_tokens}
        )

        completion = await self.client.messages.create(
            model=self.model,
            system=system_prompt,
            messages=messages,
            temperature=temperature,
            max_tokens=max_tokens,
            **kwargs
        )
        # Normalize the response to a standard dictionary format
        return {
            "choices": [{"message": {"role": "assistant", "content": completion.content[0].text}}]
        }


class Gemini(LLM):
    """LLM provider for Google (Gemini) models."""

    def __init__(self, api_key: Optional[str] = None, model: str = "gemini-1.5-flash"):
        resolved_api_key = api_key or os.getenv("GOOGLE_API_KEY")
        if not resolved_api_key:
            raise ValueError("GOOGLE_API_KEY must be provided or set as an environment variable.")

        super().__init__(api_key=resolved_api_key, model=model)
        genai.configure(api_key=self.api_key)
        self.client = genai.GenerativeModel(
            model_name=self.model,
            # System instructions are handled differently in Gemini
        )

    async def chat_completion(
        self,
        messages: List[Dict[str, str]],
        temperature: float,
        max_tokens: int,
        **kwargs: Any
    ) -> Dict[str, Any]:
        """Generates a chat completion response using the Google Gemini API."""
        # Gemini uses a different format for history and system prompts.
        # We adapt the standard message format to Gemini's format.
        system_prompt = None
        if messages and messages[0]['role'] == 'system':
            system_prompt = messages[0]['content']
            messages = messages[1:]

        # The Gemini API expects a specific format for conversation history.
        # We'll use a chat session for simplicity.
        chat = self.client.start_chat(history=[])
        if system_prompt:
            # Re-initialize the model with the system instruction
            self.client = genai.GenerativeModel(
                model_name=self.model,
                system_instruction=system_prompt
            )

        # The last message is the one we want a response for.
        user_prompt = messages[-1]['content']
        history = messages[:-1]

        # Gemini's `start_chat` history needs to be in its own format.
        # For this implementation, we'll just send the last prompt.
        # A more robust solution would convert the entire message history.
        if len(history) > 0:
            warnings.warn("Gemini provider in this implementation does not fully support history conversion yet. Only the last message will be sent.")

        logger.info(
            "Calling Gemini API.",
            extra={"model": self.model, "temperature": temperature, "max_tokens": max_tokens}
        )

        completion = await chat.send_message_async(
            user_prompt,
            generation_config=genai.types.GenerationConfig(
                max_output_tokens=max_tokens,
                temperature=temperature
            )
        )
        # Normalize the response to a standard dictionary format
        return {
            "choices": [{"message": {"role": "assistant", "content": completion.text}}]
        }

# --- Provider Factory ---

PROVIDERS: Dict[str, Type[LLM]] = {
    "openai": OpenAIProvider,
    "xai": Grok,
    "groq": GroqProvider,
    "claude": Claude,
    "gemini": Gemini,
}

def get_llm(provider_name: str, **kwargs: Any) -> LLM:
    """Factory function to get an instance of an LLM provider."""
    provider_class = PROVIDERS.get(provider_name.lower())
    if not provider_class:
        raise ValueError(f"Unknown LLM provider: '{provider_name}'. Available: {list(PROVIDERS.keys())}")
    return provider_class(**kwargs)